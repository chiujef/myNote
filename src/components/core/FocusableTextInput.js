import PropTypes from 'prop-types';
import React from 'react';
import {
  StyleSheet,
  Text,
  TextInput,
  TouchableWithoutFeedback,
  View,
} from 'react-native';

export default class FocusableTextInput extends React.Component {
  state = {
    isFocus: this.props.isFocus || false,
  };

  onPress = () => {
    this.setState({isFocus: true});
    this.props.onPress && this.props.onPress();
  };

  onBlur = () => {
    this.setState({isFocus: false});
    this.props.onBlur && this.props.onBlur();
  };

  render() {
    const {text, textInputStyle, onChangeText} = this.props;
    const {isFocus} = this.state;

    return (
      <View style={styles.container}>
        {isFocus ? (
          <TextInput
            style={[styles.textInput, textInputStyle]}
            multiline={true}
            onChangeText={onChangeText}
            onBlur={this.onBlur}
            blurOnSubmit={false}
            value={text}
            autoCorrect={false}
            autoFocus={true}
            autoCompleteType="off"
          />
        ) : (
          <TouchableWithoutFeedback style={styles.text} onPress={this.onPress}>
            <Text style={textInputStyle}>{text}</Text>
          </TouchableWithoutFeedback>
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});

FocusableTextInput.propTypes = {
  text: PropTypes.string,
  isFocus: PropTypes.bool,
  onChangeText: PropTypes.func,
  onPress: PropTypes.func,
  onBlur: PropTypes.func,
};

/*export default FocusableTextInput;*/
