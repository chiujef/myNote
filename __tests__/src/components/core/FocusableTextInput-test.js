import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {render, fireEvent} from 'react-native-testing-library';

import FocusableTextInput from '@components/core/FocusableTextInput';

describe('FocusableTextInput', () => {
  describe('Rendering', () => {
    describe('Not focused component', () => {
      it('should match to snapshot', () => {
        const component = renderer
          .create(<FocusableTextInput text="text" isFocus={false} />)
          .toJSON();
        expect(component).toMatchSnapshot();
      });
    });

    describe('Focused component', () => {
      it('should match to snapshot', () => {
        const component = renderer
          .create(<FocusableTextInput text="text" isFocus={true} />)
          .toJSON();
        expect(component).toMatchSnapshot();
      });
    });
  });

  describe('Pressing text on unfocused component', () => {
    let renderResult;
    let onPressHandler = jest.fn();

    beforeEach(() => {
      renderResult = render(
        <FocusableTextInput text="Test" onPress={onPressHandler} />,
      );
    });

    it('should trigger onPress event', () => {
      const {getByText} = renderResult;
      fireEvent(getByText('Test'), 'onPress');

      expect(onPressHandler).toHaveBeenCalled();
    });

    it('should show the TextInput', () => {
      const {getByText, getByType} = renderResult;
      fireEvent(getByText('Test'), 'onPress');

      expect(getByType('TextInput')).not.toBeNull();
    });
  });

  describe('Changing text on focused component', () => {
    let renderResult;
    let onChangeTextHandler = jest.fn();

    beforeEach(() => {
      renderResult = render(
        <FocusableTextInput
          text="Test"
          isFocus={true}
          onChangeText={onChangeTextHandler}
        />,
      );
    });

    it('should trigger onChangeText event', () => {
      const {getByType} = renderResult;
      fireEvent.changeText(getByType('TextInput'), 'ab');

      expect(onChangeTextHandler).toHaveBeenCalled();
    });
  });

  describe('Unfocusing TextInput on focused component', () => {
    let renderResult;
    let onBlurHandler = jest.fn();

    beforeEach(() => {
      renderResult = render(
        <FocusableTextInput
          text="Test"
          isFocus={true}
          onBlur={onBlurHandler}
        />,
      );
    });

    it('should trigger onBlur event', () => {
      const {getByType} = renderResult;
      fireEvent(getByType('TextInput'), 'onBlur');

      expect(onBlurHandler).toHaveBeenCalled();
    });

    it('should hide TextInput', () => {
      const {queryByType} = renderResult;
      fireEvent(queryByType('TextInput'), 'onBlur');

      expect(queryByType('TextInput')).toBeNull();
    });
  });
});
