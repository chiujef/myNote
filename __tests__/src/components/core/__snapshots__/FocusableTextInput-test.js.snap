// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusableTextInput Rendering Focused component should match to snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCompleteType="off"
    autoCorrect={false}
    autoFocus={true}
    blurOnSubmit={false}
    multiline={true}
    onBlur={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        undefined,
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value="text"
  />
</View>
`;

exports[`FocusableTextInput Rendering Not focused component should match to snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    accessible={true}
    clickable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    text
  </Text>
</View>
`;
